test/integration/virtual-rules/aria-allowed-role.js

```diff
@@ -0,0 +1,106 @@
+describe('aria-allowed-role virtual-rule', function() {
+  afterEach(function() {
+    axe.reset();
+  });
+
+  it('should pass for allowed role', function() {
+    var results = axe.runVirtualRule('aria-allowed-role', {
+      nodeName: 'div',
+      attributes: {
+        role: 'checkbox',
+        'aria-checked': true
+      }
+    });
+
+    assert.lengthOf(results.passes, 1);
+    assert.lengthOf(results.violations, 0);
+    assert.lengthOf(results.incomplete, 0);
+  });
+
+  it('should pass for element with no role', function() {
+    var results = axe.runVirtualRule('aria-allowed-attr', {
+      nodeName: 'div',
+      attributes: {
+        'aria-checked': true
+      }
+    });
+
+    assert.lengthOf(results.passes, 1);
+    assert.lengthOf(results.violations, 0);
+    assert.lengthOf(results.incomplete, 0);
+  });
+
+  it('should fail for unallowed role', function() {
+    var results = axe.runVirtualRule('aria-allowed-role', {
+      nodeName: 'dd',
+      attributes: {
+        role: 'link'
+      }
+    });
+
+    assert.lengthOf(results.passes, 0);
+    assert.lengthOf(results.violations, 1);
+    assert.lengthOf(results.incomplete, 0);
+  });
+
+  it('should pass for element with ignored option', function() {
+    axe.configure({
+      checks: [
+        {
+          id: 'aria-allowed-role',
+          options: {
+            ignoredTags: ['dd']
+          }
+        }
+      ]
+    });
+
+    var results = axe.runVirtualRule('aria-allowed-role', {
+      nodeName: 'dd',
+      attributes: {
+        role: 'link'
+      }
+    });
+
+    assert.lengthOf(results.passes, 1);
+    assert.lengthOf(results.violations, 0);
+    assert.lengthOf(results.incomplete, 0);
+  });
+
+  it('should incomplete for hidden element', function() {
```

Comment:
Why is this incomplete?

Reply:
Because hidden elements with unallowed roles return undefined https://github.com/dequelabs/axe-core/blob/develop/lib/checks/aria/aria-allowed-role-evaluate.js#L33-L35